ath79: use nvmem for wrong 3e0 cal size

neheb opened the pull request:
These three devices use AR9287 chips, which have a calibration size of 440.

ping @DragonBluep @robimarko @hauke

neheb committed:
ath79: wr2543: add compatible string for wifi

Makes it clear that the calibration size is correct as most ar72xx
devices use older wifi chips.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17278
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

neheb committed:
ath79: use nvmem for wrong 3e0 cal size

These three devices use AR9287 chips, which have a calibration size of 3d8.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17278
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep commented:
I did a further investigation with this patch. These offset value can be found on:
https://elixir.bootlin.com/linux/v6.12.5/source/drivers/net/wireless/ath/ath9k/ar9003_eeprom.c#L3390
https://elixir.bootlin.com/linux/v6.12.5/source/drivers/net/wireless/ath/ath9k/eeprom_9287.c#L42
https://elixir.bootlin.com/linux/v6.12.5/source/drivers/net/wireless/ath/ath9k/eeprom_4k.c#L41
https://elixir.bootlin.com/linux/v6.12.5/source/drivers/net/wireless/ath/ath9k/eeprom_def.c#L100

```
--- a/drivers/net/wireless/ath/ath9k/eeprom.c
+++ b/drivers/net/wireless/ath/ath9k/eeprom.c
@@ -670,6 +670,11 @@ void ath9k_hw_get_gain_boundaries_pdadcs
 
 int ath9k_hw_eeprom_init(struct ath_hw *ah)
 {
+	pr_err("size of eep_ar9300 %lx\n", round_up(sizeof(struct ar9300_eeprom), 2));
+	pr_err("size of eep_ar9287 %lx\n", round_up(sizeof(struct ar9287_eeprom) + AR9287_EEP_START_LOC*2, 2));
+	pr_err("size of eep_4k %lx\n", round_up(sizeof(struct ar5416_eeprom_4k) + 64*2, 2));
+	pr_err("size of eep_def %lx\n", round_up(sizeof(struct ar5416_eeprom_def) + 0x100*2, 2));
+
 	if (AR_SREV_9300_20_OR_LATER(ah))
 		ah->eep_ops = &eep_ar9300_ops;
 	else if (AR_SREV_9287(ah)) {
```
And the log output:
```
Mon Dec 16 10:37:18 2024 kern.err kernel: [   18.472279] size of eep_ar9300 440
Mon Dec 16 10:37:18 2024 kern.err kernel: [   18.475774] size of eep_ar9287 3d8
Mon Dec 16 10:37:18 2024 kern.err kernel: [   18.479296] size of eep_4k 1f8
Mon Dec 16 10:37:18 2024 kern.err kernel: [   18.482404] size of eep_def eb8
```

Therefore, the accurate eeprom sizes are:

|older|AR9271&&AR9285|AR9287|newer|
|-|-|-|-|
|0xeb8|0x1f8|0x3d8|0x440|

However, there is another eeprom size check.
https://elixir.bootlin.com/linux/v6.12.5/source/drivers/net/wireless/ath/ath9k/init.c#L605
So the minimum eeprom size (for AR9271&&AR9285) should be `0x200`.

neheb commented:
Interesting to say the least.

neheb committed:
bmips: dsl-2750b-b1: fix calibration size

AR9287 uses 3d8, not 440 like newer chips.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17278
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

neheb commented:
Updated and added a bmips device. ping @Noltari 

Devices which should also have this change are ar7241_tplink_tl-mr3420-v1 and ar7241_tplink_tl-wr841-v7. These currently are handled in userspace and should have their own PR. Probably.

Noltari commented:
> Updated and added a bmips device. ping @Noltari

@neheb I don't own this device, but @a-gave might be able to answer this since he added support for this device:
https://github.com/openwrt/openwrt/commit/ba9c269e1e6ecfa8fc5bdc4102330755edad66c0

neheb committed:
ath79: fix calibration size for AR9287

These devices use AR9287, which uses 3d8 as the calibration size, not
440 like newer chips do. Add a compatible line to make it clear that
this is the case.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17278
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

a-gave commented:
> > Updated and added a bmips device. ping @Noltari
> 
> @neheb I don't own this device, but @a-gave might be able to answer this since he added support for this device: [ba9c269](https://github.com/openwrt/openwrt/commit/ba9c269e1e6ecfa8fc5bdc4102330755edad66c0)

I checked on these two similar devices:
- dlink dsl-2640B E1 - AW4337AU
https://forum.archive.openwrt.org/viewtopic.php?id=38534 that runs without issues a openwrt-23.05.2 with the profile d-link,dsl-274xb-f1 
but it has a AR9285 and indeed it has 0x1f8 as length
<details>
<summary>hexdump -C /dev/mtd5 # of dlink dsl-2640B E1</summary>
<pre></code>
00000000  00 03 7f e0 00 3d ff ff  ff ff ff ff ff ff ff ff  |.....=..........|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00001000  a5 5a 00 00 00 03 50 00  16 8c 00 2b 50 08 00 01  |.Z....P....+P...|
00001010  02 80 50 2c 16 8c 30 a1  50 64 8c c0 05 04 57 0c  |..P,..0.Pd....W.|
00001020  3f 01 22 00 50 6c 3c 11  00 03 40 04 05 0b 00 4a  |?.".Pl<...@....J|
00001030  40 74 00 03 00 00 40 00  50 01 01 c2 51 0c 20 10  |@t....@.P...Q. .|
00001040  00 06 51 64 14 12 ff 24  51 68 17 ff 00 15 50 68  |..Qd...$Qh....Ph|
00001050  20 10 00 19 ff ff ff ff  ff ff ff ff ff ff ff ff  | ...............|
00001060  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00001080  01 78 49 6c e0 0e 02 01  00 00 00 00 84 c9 b2 f6  |.xIl............|
00001090  ef e8 01 01 00 00 00 00  00 00 00 09 1f 00 05 00  |................|
000010a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000010b0  00 00 00 00 00 00 00 00  00 00 00 00 00 2c 20 00  |............., .|
000010c0  e2 00 0d 00 02 0e 1c ff  0a 01 00 00 06 46 34 01  |.............F4.|
000010d0  0e 0e 00 00 00 2c 00 00  34 04 44 44 33 93 33 03  |.....,..4.DD3.3.|
000010e0  00 92 00 00 80 00 00 00  00 00 00 00 00 00 00 00  |................|
000010f0  00 00 00 00 00 00 00 00  70 8e ac 0d 1d 26 37 4b  |........p....&7K|
00001100  42 4d 59 65 69 08 12 1b  37 70 16 1f 30 4c 54 1b  |BMYei...7p..0LT.|
00001110  1b 2f 38 4a 40 4d 58 64  6e 0d 0e 1e 2d 58 10 19  |./8J@MXdn...-X..|
00001120  28 3b 58 1a 1a 2f 36 4a  3f 4c 59 64 70 0d 0d 1b  |(;X../6J?LYdp...|
00001130  28 51 0f 17 26 39 5e 70  22 22 22 22 b8 22 22 22  |(Q..&9^p""""."""|
00001140  22 ff 00 00 00 00 70 22  22 20 1e 89 22 22 20 1e  |".....p"" .."" .|
00001150  ac 22 22 20 1e 70 22 22  22 22 22 20 1c 1a 89 22  |."" .p""""" ..."|
00001160  22 22 22 22 20 1c 1a ac  22 22 22 22 22 20 1c 1a  |"""" ...""""" ..|
00001170  70 20 20 20 20 20 1e 1a  16 89 20 20 20 20 20 1e  |p     ....     .|
00001180  1a 16 ac 20 20 20 20 20  1e 1a 16 11 12 15 17 41  |...     .......A|
00001190  42 45 47 31 32 35 37 70  22 75 22 7a 62 a2 22 70  |BEG1257p"u"zb."p|
000011a0  1b 75 62 9d 62 a2 1d 70  19 75 62 9d 62 a2 1b 7a  |.ub.b..p.ub.b..z|
000011b0  13 7f 5b 93 1b 98 12 70  22 75 62 ac 22 b8 22 70  |..[....p"ub."."p|
000011c0  22 75 62 ac 22 00 00 70  22 75 62 ac 22 00 00 7a  |"ub."..p"ub."..z|
000011d0  20 7f 60 93 60 a2 20 70  1d 75 5d ac 1d 00 00 70  | .`.`. p.u]....p|
000011e0  1e 75 5e ac 1e 00 00 70  1d 75 5d ac 1d 00 00 7a  |.u^....p.u]....z|
000011f0  1c 7f 5c 93 5c a2 1c 00  ff ff ff ff ff ff ff ff  |..\.\...........|
00001200  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00010000
</code></pre></details>

- dlink dsl-2750b b1 - AW4339U - with ar9287
Here I lost calibration data years ago (https://forum.openwrt.org/t/wifi-ar922x-not-detected-after-upgrade-18-06/18338/23)
but hopefully i found https://github.com/gwlim/art-radio
and now it runs well with one of those (all of them for ar9287 are long 0x3d8)
I copied 0x440 from the other devices already mentioned while writing the dts for this device, 
but looking at other art backups in that repo it should be 0x3d8 actually. I can eventually test this on the real device, but it seems fine the new length.

neheb commented:
@a-gave thank you for confirming

@DragonBluep I added some more AR9287 fixes.

neheb committed:
lantiq: use nvmem for fritz73x0

These units use AR9287, which has a calibration size of 3d8.

Also fixed compatible string to the proper one to indicate AR9287.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17278
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

neheb committed:
lantiq: use nvmem for fritz 736x

Two use AR9287 and one AR9381. Both have different calibration sizes.
Move differences out of wifi node to make it clearer what's what.

qca,no-eeprom needs to stay for 7362sl as there's no nvmem equivalent
for caldata_extract_reverse

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17278
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

neheb commented:
Added lantiq AR9287 fixes. ping @xdarklight @abajk @jneuhauser

neheb commented:
Removed tp-link lantiq devices. Those are handled in a different PR.
